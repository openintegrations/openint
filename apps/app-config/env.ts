export * from '@openint/env/env'
