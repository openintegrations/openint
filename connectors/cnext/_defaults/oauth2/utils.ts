import {z} from 'zod'
import {zOAuthConfig, zTokenResponse} from './def'

export function prepareScopes(jsonConfig: z.infer<typeof zOAuthConfig>) {
  const scopes = jsonConfig.scopes
  const scopeSeparator = jsonConfig.scope_separator
  return encodeURIComponent(
    scopes.map((s) => s.scope).join(scopeSeparator ?? ' '),
  )
}

/**
 * Fills out variables in a URL string template with values from connector config and connection settings
 *
 * @param url - URL string containing variables in ${variable} format
 * @param connectorConfig - Connector configuration object containing values to substitute
 * @param connectionSettings - Connection settings object containing values to substitute
 * @returns URL string with variables replaced with actual values
 *
 * Variables can reference:
 * - Connector config values using ${connector_config.key}
 * - Connection settings using ${connection_settings.key}
 *
 * Example:
 * Input: "https://api.example.com/${connector_config.version}/users/${connection_settings.user_id}"
 * Output: "https://api.example.com/v1/users/123"
 */

export function fillOutStringTemplateVariables(
  url: string,
  connectorConfig: any,
  connectionSettings: any,
) {
  if (!url) return url
  let filledUrl = url

  // Ensure connectorConfig and connectionSettings are objects
  connectorConfig = connectorConfig || {}
  connectionSettings = connectionSettings || {}

  // Match ${variable} pattern
  const variableRegex = /\${([^}]+)}/g
  const matches = url.match(variableRegex)

  if (!matches) {
    return filledUrl
  }

  matches.forEach((match) => {
    // Extract variable name without ${} wrapper
    const varName = match.slice(2, -1)

    // Check if variable references connector_config
    if (varName.startsWith('connector_config.')) {
      const configKey = varName.split('.')[1]
      const value = connectorConfig[configKey as keyof typeof connectorConfig]
      if (value) {
        filledUrl = filledUrl.replace(match, value)
      }
    }
    // Check if variable references connection_settings
    else if (varName.startsWith('connection_settings.')) {
      const settingKey = varName.split('.')[1]
      const value =
        connectionSettings[settingKey as keyof typeof connectionSettings]
      if (value) {
        filledUrl = filledUrl.replace(match, value)
      }
    }
  })

  return filledUrl
}
export async function makeTokenRequest(
  url: string,
  params: Record<string, string>,
  flowType: 'exchange' | 'refresh',
  // note: we may want to add bodyFormat: form or json
): Promise<z.infer<typeof zTokenResponse>> {
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Accept: 'application/json',
    },
    body: new URLSearchParams(params),
  })

  if (!response.ok) {
    const errorText = await response.text()
    throw new Error(
      `Token ${flowType} failed: ${response.status} ${response.statusText} - ${errorText}`,
    )
  }

  try {
    return zTokenResponse.parse(await response.json())
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new Error(
        `Invalid oauth2 ${flowType} token response format: ${error.message}`,
      )
    }
    throw error
  }
}
