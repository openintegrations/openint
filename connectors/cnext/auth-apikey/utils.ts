import type {ConnectContext} from '@openint/cdk'
import type {Z} from '@openint/util/zod-utils'
import type {apiKeySchemas, zApiKeyConfig} from './schemas'

import {initGreenhouseSDK} from '@opensdks/sdk-greenhouse'

export function getClient({
  connectorName,
  oauthConfigTemplate,
  connectorConfig,
  connectionSettings,
  ...connectCtx
}: {
  connectorName: string
  oauthConfigTemplate: Z.infer<typeof zApiKeyConfig>
  connectorConfig: Z.infer<typeof apiKeySchemas.connector_config>
  connectionSettings:
    | Z.infer<typeof apiKeySchemas.connection_settings>
    | undefined
} & Pick<ConnectContext<{}>, 'baseURLs' | 'fetch'>) {
  // TODO: Update since for now this only creates a greenhouse client.
  // TODO: Replace with http calls or directly use a fetch post as we do for oauth2
  // and use that in checkConnection instead.
  const client = initGreenhouseSDK({
    auth: {
      basic: {
        username: connectionSettings?.oauth.credentials?.api_key ?? '',
      },
    },
  })

  return {
    client,
    apiKeyConfig: {
      connectorConfig,
      connectionSettings: connectionSettings ?? {},
      // TODO: We might not need this at all
      baseURLs: connectCtx.baseURLs,
    },
  }
}
