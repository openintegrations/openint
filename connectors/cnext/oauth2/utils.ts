import {type Z} from '@openint/util/zod-utils'
import {zOAuthConfig} from './def'

export function prepareScopes(
  scopes: string[],
  jsonConfig: Z.infer<typeof zOAuthConfig>,
) {
  const scopeSeparator = jsonConfig.scope_separator ?? ' '

  if (!scopes || !Array.isArray(scopes) || scopes.length === 0) {
    return ''
  }
  return scopes.join(encodeURIComponent(scopeSeparator))
}

/**
 * Fills out variables in a URL string template with values from connector config and connection settings
 *
 * @param string - URL string containing variables in ${variable} format
 * @param connectorConfig - Connector configuration object containing values to substitute
 * @param connectionSettings - Connection settings object containing values to substitute
 * @returns URL string with variables replaced with actual values
 *
 * Variables can reference:
 * - Connector config values using ${connector_config.key}
 * - Connection settings using ${connection_settings.key}
 *
 * Example:
 * Input: "https://api.example.com/${connector_config.version}/users/${connection_settings.user_id}"
 * Output: "https://api.example.com/v1/users/123"
 */

export function fillOutStringTemplateVariables(
  string: string,
  connectorConfig: any,
  connectionSettings: any,
) {
  if (!string) return string
  let filledUrl = string

  // Ensure connectorConfig and connectionSettings are objects
  connectorConfig = connectorConfig || {}
  connectionSettings = connectionSettings || {}

  // Match ${variable} pattern
  const variableRegex = /\${([^}]+)}/g
  const matches = string.match(variableRegex)

  if (!matches) {
    return filledUrl
  }

  matches.forEach((match) => {
    // Extract variable name without ${} wrapper
    const varName = match.slice(2, -1)

    // Check if variable references connector_config
    if (varName.startsWith('connector_config.')) {
      const configKey = varName.split('.')[1]
      const value = connectorConfig[configKey as keyof typeof connectorConfig]
      if (value) {
        filledUrl = filledUrl.replace(match, value)
      }
    }
    // Check if variable references connection_settings
    else if (varName.startsWith('connection_settings.')) {
      const settingKey = varName.split('.')[1]
      const value =
        connectionSettings[settingKey as keyof typeof connectionSettings]
      if (value) {
        filledUrl = filledUrl.replace(match, value)
      }
    }
  })

  return filledUrl
}

export function fillOutStringTemplateVariablesInObjectKeys(
  obj: any,
  connectorConfig: any,
  connectionSettings: any,
): any {
  if (!obj || typeof obj !== 'object') {
    return obj
  }

  // Handle arrays properly
  if (Array.isArray(obj)) {
    return obj.map((item: any) =>
      fillOutStringTemplateVariablesInObjectKeys(
        item,
        connectorConfig,
        connectionSettings,
      ),
    )
  }

  const clone = {...obj}

  for (const key of Object.keys(clone)) {
    const value = clone[key]

    if (typeof value === 'string') {
      clone[key] = fillOutStringTemplateVariables(
        value,
        connectorConfig,
        connectionSettings,
      )
    } else if (value !== null && typeof value === 'object') {
      clone[key] = fillOutStringTemplateVariablesInObjectKeys(
        value,
        connectorConfig,
        connectionSettings,
      )
    }
  }

  return clone
}

/*
 * This function takes the paramNames map where a user can map were fields like client_id and client_secret are named in particular oauth connector.
 * For example salesforce may call client_id clientKey. In this case, the paramNames would have client_id: clientKey.
 * Following the SF example, this function will return a new object with the client_id field renamed to clientKey.
 * Write tests for this function to in different scenarios ensure that the clientKey is returned with the value initially set for client_id
 */
export function mapOauthParams(
  params: Record<string, string>,
  paramNames: Record<string, string>,
) {
  const result: Record<string, string> = {}

  // Process each parameter in the input
  Object.entries(params).forEach(([key, value]) => {
    if (key && paramNames && key in paramNames && paramNames[key]) {
      result[paramNames[key]] = value
    } else {
      result[key] = value
    }
  })

  return result
}
