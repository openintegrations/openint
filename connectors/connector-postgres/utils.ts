import {sql} from 'drizzle-orm'
import type {PgColumn, PgDatabase} from 'drizzle-orm/pg-core'
import {pgTable, uniqueIndex} from 'drizzle-orm/pg-core'
import type {RecordMessageBody} from './def'

export function isValidDateString(str: string) {
  const date = new Date(str)
  const lenToCheck = Math.min(str.length, '2021-01-01T00:00:00'.length)
  return (
    !Number.isNaN(date.getTime()) &&
    date.toISOString().slice(0, lenToCheck) === str.slice(0, lenToCheck)
  )
}

/**
 * We probably want something based off of the Catalog protocol, by inferring catalog
 * from data object and then generating a migration from the catalog message as a two step,
 * unified process.
 * For now we just do something lightweig thought
 *
 * Also consider deduping with inferTableForUpsert
 */
export function inferTableFromMessage(event: RecordMessageBody) {
  // TODO Implement namespace and upsert metadata support
  return pgTable(
    event.stream,
    (t) => {
      function inferCol(v: unknown) {
        if (typeof v === 'string') {
          return isValidDateString(v) ? t.timestamp({mode: 'string'}) : t.text()
        }
        if (typeof v === 'number') {
          return Math.floor(v) === v ? t.integer() : t.doublePrecision()
        }
        if (typeof v === 'boolean') {
          return t.boolean()
        }
        if (v instanceof Date) {
          return t.timestamp()
        }
        return t.jsonb()
      }
      return Object.fromEntries(
        Object.entries(event.data).map(([k, v]) => [k, inferCol(v)]),
      )
    },
    (table) =>
      // TODO: Check against existing primary key and unique indexes
      // to not create if already there?
      // && Object.values(table).filter(c => c.i)
      event.upsert?.key_columns
        ? [
            uniqueIndex(`${event.stream}_upsert_keys`).on(
              ...(event.upsert.key_columns.map(
                (col) => table[col] as PgColumn,
              ) as [PgColumn]),
            ),
          ]
        : [],
  )
}

export async function runMigrationForStandardTable(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  db: PgDatabase<any, any>,
  tableName: string,
) {
  // TODO: turn this into a drizzle table
  // Where do we want to put data? Not always public...
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return await db.execute(sql`
    CREATE TABLE IF NOT EXISTS ${sql.identifier(tableName)} (
      source_id VARCHAR NOT NULL,
      id VARCHAR NOT NULL,
      customer_id VARCHAR,
      created_at timestamp with time zone DEFAULT now() NOT NULL,
      updated_at timestamp with time zone DEFAULT now() NOT NULL,
      connector_name VARCHAR GENERATED ALWAYS AS (split_part((source_id)::text, '_'::text, 2)) STORED NOT NULL,
      CONSTRAINT ${sql.identifier(
        `pk_${tableName}`,
      )} PRIMARY KEY ("source_id", "id"),
      unified jsonb,
      raw jsonb DEFAULT '{}'::jsonb NOT NULL
    );
  `)
  // NOTE: Should we add org_id?
  // NOTE: Rename `unified` to `unified` and `raw` to `raw` or `remote` or `original`
  // NOTE: add prefix check would be nice
  // for (const col of [
  //   'id',
  //   'source_id',
  //   'connector_name',
  //   'created_at',
  //   'updated_at',
  //   'customer_id',
  // ]) {
  //   await pool.query(sql`
  //     CREATE INDEX IF NOT EXISTS ${sql.identifier([
  //       `${tableName}_${col}`,
  //     ])} ON ${table} (${sql.identifier([col])});
  //   `)
  // }
}
