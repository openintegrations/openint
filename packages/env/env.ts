import {z} from '@opensdks/util-zod'
import {createEnv} from '@t3-oss/env-nextjs'
import {proxyRequired} from './proxyRequired'

export const envConfig = {
  server: {
    // MARK: - Not validated, may not be used...
    // Core env vars
    POSTGRES_URL: z.string().default('postgres://localhost:5432/postgres'),
    // TODO: Incorporate schema in the url itself.
    POSTGRES_SCHEMA: z.string().optional(),

    JWT_SECRET: z.string().optional(),
    CLERK_SECRET_KEY: z.string().optional(),
    NANGO_SECRET_KEY: z.string().optional(),

    // Required for worker to work when deployed
    INNGEST_SIGNING_KEY: z.string().optional(),
    INNGEST_EVENT_KEY: z.string().optional(),

    // Optional
    SENTRY_CRON_MONITOR_URL: z.string().optional(),

    // Turn on debug output, including drizzle. Should be a boolean tho
    DEBUG: z.string().optional(),

    // Variables set by Vercel when deployed
    VERCEL_URL: z.string().optional(),
    VERCEL_ENV: z.enum(['production', 'preview', 'development']).optional(),
  },
  client: {
    NEXT_PUBLIC_SERVER_URL: z.string().optional(),
    NEXT_PUBLIC_NANGO_PUBLIC_KEY: z.string().optional(),
    // Where the app is running. Only used by getServerUrl at the moment
    NEXT_PUBLIC_PORT: z.string().optional(),

    NEXT_PUBLIC_SUPABASE_URL: z.string().optional(),
    NEXT_PUBLIC_SUPABASE_ANON_KEY: z.string().optional(),
    NEXT_PUBLIC_SENTRY_DSN: z.string().optional(),
    NEXT_PUBLIC_SENTRY_ORG: z.string().optional(),
    NEXT_PUBLIC_POSTHOG_WRITEKEY: z.string().optional(),
    NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY: z.string().optional(),
    NEXT_PUBLIC_CLERK_SUPABASE_JWT_TEMPLATE_NAME: z
      .string()
      .default('supabase'),
    NEXT_PUBLIC_COMMANDBAR_ORG_ID: z.string().optional(),
  },
  runtimeEnv: overrideFromLocalStorage({
    NEXT_PUBLIC_COMMANDBAR_ORG_ID: process.env['NEXT_PUBLIC_COMMANDBAR_ORG_ID'],
    NEXT_PUBLIC_SUPABASE_URL: process.env['NEXT_PUBLIC_SUPABASE_URL'],
    NEXT_PUBLIC_SUPABASE_ANON_KEY: process.env['NEXT_PUBLIC_SUPABASE_ANON_KEY'],
    NEXT_PUBLIC_SENTRY_DSN: process.env['NEXT_PUBLIC_SENTRY_DSN'],
    NEXT_PUBLIC_SENTRY_ORG: process.env['NEXT_PUBLIC_SENTRY_ORG'],
    NEXT_PUBLIC_POSTHOG_WRITEKEY: process.env['NEXT_PUBLIC_POSTHOG_WRITEKEY'],
    NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY:
      process.env['NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY'],
    NEXT_PUBLIC_CLERK_SUPABASE_JWT_TEMPLATE_NAME:
      process.env['NEXT_PUBLIC_CLERK_SUPABASE_JWT_TEMPLATE_NAME'],
    DEBUG: process.env['DEBUG'],
    JWT_SECRET: process.env['JWT_SECRET'],
    CLERK_SECRET_KEY: process.env['CLERK_SECRET_KEY'],
    NANGO_SECRET_KEY: process.env['NANGO_SECRET_KEY'],
    SENTRY_CRON_MONITOR_URL: process.env['SENTRY_CRON_MONITOR_URL'],
    VERCEL_ENV: process.env['VERCEL_ENV'],
    POSTGRES_SCHEMA: process.env['POSTGRES_SCHEMA'],
    INNGEST_EVENT_KEY: process.env['INNGEST_EVENT_KEY'],
    INNGEST_SIGNING_KEY: process.env['INNGEST_SIGNING_KEY'],
    NEXT_PUBLIC_NANGO_PUBLIC_KEY: process.env['NEXT_PUBLIC_NANGO_PUBLIC_KEY'],
    NEXT_PUBLIC_PORT: process.env['NEXT_PUBLIC_PORT'],
    NEXT_PUBLIC_SERVER_URL: process.env['NEXT_PUBLIC_SERVER_URL'],
    POSTGRES_URL: process.env['POSTGRES_URL'],
    VERCEL_URL: process.env['VERCEL_URL'],
  }),
} satisfies Parameters<typeof createEnv>[0]

export const env = createEnv(envConfig)

export const envRequired = proxyRequired(env, {
  formatError(key) {
    return new Error(`Missing required env var: ${key}`)
  },
})

export type Env = typeof env

// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
;(globalThis as any).env = env

/** Allow NEXT_PUBLIC values to be overwriten from localStorage for debugging purposes */
function overrideFromLocalStorage<T>(runtimeEnv: T) {
  if (typeof window !== 'undefined' && window.localStorage) {
    for (const key in runtimeEnv) {
      if (key.startsWith('NEXT_PUBLIC_')) {
        const value = window.localStorage.getItem(key)
        if (value != null) {
          runtimeEnv[key] = value as T[Extract<keyof T, string>]
          console.warn(`[env] Overriding from localStorage ${key} = ${value}`)
        }
      }
    }
  }
  return runtimeEnv
}
