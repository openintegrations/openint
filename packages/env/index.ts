import {z} from '@opensdks/util-zod'
import {createEnv} from '@t3-oss/env-nextjs'

export * from './env'
export * from './env-test'
export * from './proxyRequired'

/** @deprecated */
export function getEnv<T extends z.ZodTypeAny = z.ZodString>(
  name: string,
  schema?: T,
): z.infer<T> {
  const env = createEnv({
    server: {[name]: schema ?? z.string()},
    runtimeEnv: process.env,
  })
  return env[name as keyof typeof env]
}
