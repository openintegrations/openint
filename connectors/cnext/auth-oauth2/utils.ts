import type {ConnectContext} from '@openint/cdk'
import type {oauth2Schemas, zOAuthConfig} from './schemas'

import {getConnectorDefaultCredentials} from '@openint/env'
import {createOAuth2Client} from '@openint/oauth2/createOAuth2Client'
import {type Z} from '@openint/util/zod-utils'
import {renderTemplateObject} from '../lib/template'

export function injectCcfgDefaultCredentialsIfNeeded(
  connectorConfig: Z.infer<typeof oauth2Schemas.connector_config>,
  connectorName: string,
  oauthConfig: Z.infer<typeof zOAuthConfig>,
): Z.infer<typeof oauth2Schemas.connector_config> {
  const defaultCredentials = getConnectorDefaultCredentials(connectorName)
  if (
    !connectorConfig.oauth?.client_id ||
    !connectorConfig.oauth?.client_secret
  ) {
    const client_id = defaultCredentials?.['client_id']
    const client_secret = defaultCredentials?.['client_secret']
    if (!client_id || !client_secret) {
      throw new Error(
        `Missing default credentials for connector ${connectorName}`,
      )
    }

    const configuredScopes = connectorConfig.oauth?.scopes ?? []

    if (
      oauthConfig.openint_scopes &&
      configuredScopes.length > 0 &&
      !configuredScopes.every((scope) =>
        oauthConfig.openint_scopes?.includes(scope),
      )
    ) {
      const invalidScopes = configuredScopes.filter(
        (scope) => !oauthConfig.openint_scopes?.includes(scope),
      )
      throw new Error(
        `Invalid scopes configured: ${invalidScopes.join(', ')}. ` +
          `Valid default scopes are: ${oauthConfig.openint_scopes?.join(', ')}`,
      )
    }

    return {
      ...connectorConfig,
      oauth: {
        client_id,
        client_secret,
        scopes: configuredScopes ?? [],
      },
    }
  }
  return connectorConfig
}

export function getClient({
  connectorName,
  oauthConfigTemplate,
  connectorConfig,
  connectionSettings,
  ...connectCtx
}: {
  connectorName: string
  oauthConfigTemplate: Z.infer<typeof zOAuthConfig>
  connectorConfig: Z.infer<typeof oauth2Schemas.connector_config>
  connectionSettings:
    | Z.infer<typeof oauth2Schemas.connection_settings>
    | undefined
} & Pick<ConnectContext<{}>, 'baseURLs' | 'fetch'>) {
  const oauthConfig = renderTemplateObject(oauthConfigTemplate, {
    connectorConfig,
    connectionSettings: connectionSettings ?? {},
    baseURLs: connectCtx.baseURLs,
  })

  const ccfg = injectCcfgDefaultCredentialsIfNeeded(
    connectorConfig,
    connectorName,
    oauthConfig,
  )
  const client = createOAuth2Client(
    {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      clientId: ccfg.oauth!.client_id!,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      clientSecret: ccfg.oauth!.client_secret!,
      authorizeURL: oauthConfig.authorization_request_url,
      tokenURL: oauthConfig.token_request_url,
      revokeUrl: oauthConfig.revocation_request_url,
      scopeDelimiter: oauthConfig.scope_separator,
      paramKeyMapping: oauthConfig.params_config.param_names,
      clientAuthLocation: 'body', // Make this configurable
    },
    connectCtx.fetch,
  )
  return {client, oauthConfig}
}
