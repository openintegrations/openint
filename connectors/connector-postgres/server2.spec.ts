import {getTableName} from 'drizzle-orm'
import {drizzle} from 'drizzle-orm/postgres-js'
import * as R from 'remeda'
import type {EndUserId} from '@openint/cdk'
import {env} from '@openint/env'
import {rxjs, toCompletion} from '@openint/util'
import type {RecordMessageBody} from './def'
import postgresServer from './server2'
import {getMigrationsForTable, inferTable} from './utils'

beforeAll(async () => {
  const masterDb = drizzle(env.POSTGRES_URL, {logger: true})
  await masterDb.execute('DROP DATABASE IF EXISTS testing')
  await masterDb.execute('CREATE DATABASE testing')
  await masterDb.$client.end()
})

const dbUrl = new URL(env.POSTGRES_URL)
dbUrl.pathname = '/testing'
const db = drizzle(dbUrl.toString(), {logger: true})

const destLink = postgresServer.destinationSync({
  config: {},
  endUser: {id: 'esur_12' as EndUserId, orgId: 'org_123'},
  settings: {databaseUrl: dbUrl.toString()},
  source: {id: 'reso_sfdc_9287', connectorName: 'sfdc'},
  state: {},
})

describe('standard schema', () => {
  test('destinationSync', async () => {
    const src = rxjs.from([
      {
        data: {
          entityName: 'vendor',
          id: 'vend_1123',
          entity: {
            raw: {AccountName: 'Cash'},
            unified: {name: 'Assets:Cash'},
            end_user_id: 'esur_12',
            id: 'vend_1123',
            source_id: 'reso_sfdc_9287',
          },
        },
        type: 'data' as const,
      },
      {type: 'commit' as const},
    ])
    await toCompletion(destLink(src))
    const res = await db.execute('SELECT * FROM vendor')
    expect(res[0]).toMatchObject({
      source_id: 'reso_sfdc_9287',
      id: 'vend_1123',
      end_user_id: 'esur_12',
      connector_name: 'sfdc',
      unified: {name: 'Assets:Cash'},
      raw: {AccountName: 'Cash'},
      // Should be any ISODate
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      created_at: expect.any(String),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      updated_at: expect.any(String),
    })
  })
})

const messages = [
  ['insert', [{entityName: 'house', entity: {id: 112, name: 'White'}}]],
  [
    'insert multiple',
    [
      {entityName: 'account', entity: {id: 555, name: 'Bank'}},
      {entityName: 'account', entity: {id: 555}},
    ],
  ],
  [
    'upsert single key',
    [
      {
        entityName: 'tre',
        entity: {id: 5, name: 'B'},
        upsert: {key_columns: ['id']},
      },
    ],
  ],
  [
    'upsert composite keys',
    [
      {
        entityName: 'transaction',
        entity: {
          id: '1',
          intId: 2,
          amount: 1.23,
          is_deleted: false,
          raw: {key: 'value'},
          array_value: [{key: 'value'}],
          null_value: null,
          undefined_value: undefined, // undefined gets turned into null... but should we allowe for it?
          // created_at: '2021-01-01T00:00:00Z', // this colun also causes problem for now
          // uupdated_at: new Date(), // Should we allow non-json values?
        },
        upsert: {
          key_columns: ['id', 'is_deleted'],
        },
      },
    ],
  ],
  // [
  //   'upsert multiple single key',
  //   [
  //     {entityName: 'tre', entity: {id: 5, name: 'B'}, upsert: {key_columns: ['id']}},
  //     // Needs to be two batches..n in order to work due to postgres upsert cannot operate on the same row twice in one txn
  //     {entityName: 'tre', entity: {id: 5, name: 'B'}, upsert: {key_columns: ['id']}},
  //   ],
  // ],
  [
    'mixed',
    [
      {
        entityName: 'good',
        entity: {id: 5, name: 'B'},
        upsert: {key_columns: ['id']},
      },
      {
        entityName: 'life',
        entity: {id: 1, code: 'C'},
        upsert: {key_columns: ['id']},
      },
    ],
  ],
] satisfies Array<[string, Array<Omit<RecordMessageBody, 'id'>>]>

describe.each(messages)('postgresServer: %s', (_, messages) => {
  const tables = messages.map((m) =>
    inferTable(m as unknown as RecordMessageBody),
  )

  test('infer table schema', async () => {
    const tableMigrations = await Promise.all(
      tables.map((t) => getMigrationsForTable(t)),
    )
    expect(tableMigrations.flat()).toMatchSnapshot()
    for (const migration of tableMigrations.flat()) {
      await db.execute(migration)
    }
  })

  test('destinationSync', async () => {
    const src = rxjs.from([
      ...messages.map((m) => ({
        data: {id: '', ...m},
        type: 'data' as const,
      })),
      {type: 'commit' as const},
    ])

    await toCompletion(destLink(src))

    const uniqueTables = R.uniqBy(tables, (t) => getTableName(t))
    console.log({
      uniqueTables: uniqueTables.length,
      names: tables.map((t) => getTableName(t)),
    })

    const tableRows = await Promise.all(
      uniqueTables.map((t) => db.select().from(t).execute()),
    )

    const rowsByTable = Object.fromEntries(
      uniqueTables.map((t, i) => [getTableName(t), tableRows[i]]),
    )

    expect(rowsByTable).toMatchSnapshot()

    // for (const [index, message] of messages.entries()) {
    //   // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    //   const table = tables[index]!
    //   // This wont' work for multiple.. in same table.. need to see.
    //   const rows = await db.select().from(table).execute()
    //   // expect(rows[0]).toEqual(message.data)
    // }
  })
})

afterAll(() => db.$client.end())
