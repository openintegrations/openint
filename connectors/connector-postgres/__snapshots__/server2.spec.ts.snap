// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postgresServer: insert destinationSync 1`] = `
{
  "house": [
    {
      "id": 112,
      "name": "White",
    },
  ],
}
`;

exports[`postgresServer: insert infer table schema 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "house" (
	"id" integer,
	"name" text
);
",
]
`;

exports[`postgresServer: insert multiple destinationSync 1`] = `
{
  "account": [
    {
      "id": 555,
      "name": "Bank",
    },
    {
      "id": 555,
      "name": null,
    },
  ],
}
`;

exports[`postgresServer: insert multiple infer table schema 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "account" (
	"id" integer,
	"name" text
);
",
  "CREATE TABLE IF NOT EXISTS "account" (
	"id" integer
);
",
]
`;

exports[`postgresServer: mixed destinationSync 1`] = `
{
  "good": [
    {
      "id": 5,
      "name": "B",
    },
  ],
  "life": [
    {
      "code": "C",
      "id": 1,
    },
  ],
}
`;

exports[`postgresServer: mixed infer table schema 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "good" (
	"id" integer,
	"name" text
);
",
  "CREATE UNIQUE INDEX IF NOT EXISTS "good_upsert_keys" ON "good" USING btree ("id");",
  "CREATE TABLE IF NOT EXISTS "life" (
	"id" integer,
	"code" text
);
",
  "CREATE UNIQUE INDEX IF NOT EXISTS "life_upsert_keys" ON "life" USING btree ("id");",
]
`;

exports[`postgresServer: upsert composite keys destinationSync 1`] = `
{
  "transaction": [
    {
      "amount": 1.23,
      "array_value": [
        {
          "key": "value",
        },
      ],
      "id": "1",
      "intId": 2,
      "is_deleted": false,
      "null_value": null,
      "raw": {
        "key": "value",
      },
      "undefined_value": null,
    },
  ],
}
`;

exports[`postgresServer: upsert composite keys infer table schema 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "transaction" (
	"id" text,
	"intId" integer,
	"amount" double precision,
	"is_deleted" boolean,
	"raw" jsonb,
	"array_value" jsonb,
	"null_value" jsonb,
	"undefined_value" jsonb
);
",
  "CREATE UNIQUE INDEX IF NOT EXISTS "transaction_upsert_keys" ON "transaction" USING btree ("id","is_deleted");",
]
`;

exports[`postgresServer: upsert single key destinationSync 1`] = `
{
  "tre": [
    {
      "id": 5,
      "name": "B",
    },
  ],
}
`;

exports[`postgresServer: upsert single key infer table schema 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "tre" (
	"id" integer,
	"name" text
);
",
  "CREATE UNIQUE INDEX IF NOT EXISTS "tre_upsert_keys" ON "tre" USING btree ("id");",
]
`;
